dmypy: INTERNAL ERROR: AssertionError in _add_error_info()

## Crash Report

Just trying out dmypy for the first time and I encountered the below issue when trying to check an individual file after starting the daemon and checking the entire source repository.

## Repro / Traceback
```
$ time dmypy run -- src/ ; time dmypy run -- src/scratch.py ; time dmypy run -- src/scratch.py
Daemon started
src/scratch.py:16: error: "C" has no attribute "foo"  [attr-defined]
Found 1 error in 1 file (checked 2952 source files)

real    2m1.636s
user    0m0.264s
sys     0m0.140s
src/scratch.py:16: error: "C" has no attribute "foo"  [attr-defined]
Found 1 error in 1 file (checked 1 source file)

real    0m54.281s
user    0m0.174s
sys     0m0.098s
-------------------------------------------------------------------------------
stderr:
src/scratch.py:16: error: INTERNAL ERROR -- Please try using mypy master on GitHub:
https://mypy.readthedocs.io/en/stable/common_issues.html#using-a-development-mypy-build
Please report a bug at https://github.com/python/mypy/issues
version: 1.3.0
src/scratch.py:16: error: INTERNAL ERROR -- Please try using mypy master on GitHub:
https://mypy.readthedocs.io/en/stable/common_issues.html#using-a-development-mypy-build
Please report a bug at https://github.com/python/mypy/issues
version: 1.3.0
Daemon crashed!
Traceback (most recent call last):
  File "mypy/dmypy_server.py", line 230, in serve
  File "mypy/dmypy_server.py", line 277, in run_command
  File "mypy/dmypy_server.py", line 345, in cmd_run
  File "mypy/dmypy_server.py", line 414, in check
  File "mypy/dmypy_server.py", line 663, in fine_grained_increment_follow_imports
  File "mypy/server/update.py", line 288, in update
  File "mypy/server/update.py", line 881, in propagate_changes_using_dependencies
  File "mypy/server/update.py", line 1024, in reprocess_nodes
  File "mypy/checker.py", line 526, in check_second_pass
  File "mypy/checker.py", line 531, in check_partial
  File "mypy/checker.py", line 545, in check_top_level
  File "mypy/nodes.py", line 1228, in accept
  File "mypy/checker.py", line 4130, in visit_expression_stmt
  File "mypy/checkexpr.py", line 4892, in accept
  File "mypy/errors.py", line 1177, in report_internal_error
  File "mypy/checkexpr.py", line 4882, in accept
  File "mypy/checkexpr.py", line 429, in visit_call_expr
  File "mypy/checkexpr.py", line 515, in visit_call_expr_inner
  File "mypy/checkexpr.py", line 4892, in accept
  File "mypy/errors.py", line 1177, in report_internal_error
  File "mypy/checkexpr.py", line 4890, in accept
  File "mypy/nodes.py", line 1813, in accept
  File "mypy/checkexpr.py", line 2771, in visit_member_expr
  File "mypy/checkexpr.py", line 2792, in analyze_ordinary_member_access
  File "mypy/checkmember.py", line 201, in analyze_member_access
  File "mypy/checkmember.py", line 220, in _analyze_member_access
  File "mypy/checkmember.py", line 343, in analyze_instance_member_access
  File "mypy/checkmember.py", line 601, in analyze_member_var_access
  File "mypy/checkmember.py", line 274, in report_missing_attribute
  File "mypy/messages.py", line 485, in has_no_attr
  File "mypy/messages.py", line 272, in fail
  File "mypy/messages.py", line 247, in report
  File "mypy/errors.py", line 440, in report
  File "mypy/errors.py", line 502, in add_error_info
  File "mypy/errors.py", line 443, in _add_error_info
AssertionError


real    0m0.863s
user    0m0.158s
sys     0m0.107s
```

## Observations
1. Checking our entire source repo of 2,952 files takes ~2 minutes, and then re-checking a single file (which has not been modified, was already checked, and had its errors reported in the first run) takes nearly a minute. This seems very surprising. (I'd expect this to be near instant, given the file has not been modified since the initial check?)
2. This crash only seems to happen if `scratch.py` has an error it it. If I 'fix' the error in that file, the crash does not occur, though the curiosity of re-checking the individual file taking nearly a minute remains.

## Environment
- Mypy version used: 1.3.0
- Mypy configuration options from `mypy.ini` (and other config files): _Can repro with empty mypy.ini_
- Python version used: 3.11.3
- Operating system and version: Red Hat Enterprise Linux release 8.3 (Ootpa)

I'm seeing a very similar stacktrace after upgrading to the latest mypy 1.14.1:

```
src/stellarpunk/orders/core.py:686: error: INTERNAL ERROR -- Please try using mypy master on GitHub:
https://mypy.readthedocs.io/en/stable/common_issues.html#using-a-development-mypy-build
Please report a bug at https://github.com/python/mypy/issues
version: 1.14.1
Daemon crashed!
Traceback (most recent call last):
  File "mypy/dmypy_server.py", line 236, in serve
  File "mypy/dmypy_server.py", line 285, in run_command
  File "mypy/dmypy_server.py", line 353, in cmd_run
  File "mypy/dmypy_server.py", line 432, in check
  File "mypy/dmypy_server.py", line 700, in fine_grained_increment_follow_imports
  File "mypy/server/update.py", line 285, in update
  File "mypy/server/update.py", line 881, in propagate_changes_using_dependencies
  File "mypy/server/update.py", line 1025, in reprocess_nodes
  File "mypy/checker.py", line 535, in check_second_pass
  File "mypy/checker.py", line 547, in check_partial
  File "mypy/checker.py", line 594, in accept
  File "mypy/errors.py", line 1288, in report_internal_error
  File "mypy/checker.py", line 592, in accept
  File "mypy/nodes.py", line 827, in accept
  File "mypy/checker.py", line 1046, in visit_func_def
  File "mypy/checker.py", line 1050, in _visit_func_def
  File "mypy/checker.py", line 1084, in check_func_item
  File "mypy/checker.py", line 1360, in check_func_def
  File "mypy/checker.py", line 594, in accept
  File "mypy/errors.py", line 1288, in report_internal_error
  File "mypy/checker.py", line 592, in accept
  File "mypy/nodes.py", line 1277, in accept
  File "mypy/checker.py", line 2952, in visit_block
  File "mypy/checker.py", line 594, in accept
  File "mypy/errors.py", line 1288, in report_internal_error
  File "mypy/checker.py", line 592, in accept
  File "mypy/nodes.py", line 1473, in accept
  File "mypy/checker.py", line 4613, in visit_return_stmt
  File "mypy/checker.py", line 4649, in check_return_stmt
  File "mypy/checkexpr.py", line 5892, in accept
  File "mypy/errors.py", line 1288, in report_internal_error
  File "mypy/checkexpr.py", line 5890, in accept
  File "mypy/nodes.py", line 1908, in accept
  File "mypy/checkexpr.py", line 3283, in visit_member_expr
  File "mypy/checkexpr.py", line 3308, in analyze_ordinary_member_access
  File "mypy/checkmember.py", line 206, in analyze_member_access
  File "mypy/checkmember.py", line 225, in _analyze_member_access
  File "mypy/checkmember.py", line 356, in analyze_instance_member_access
  File "mypy/checkmember.py", line 618, in analyze_member_var_access
  File "mypy/checkmember.py", line 279, in report_missing_attribute
  File "mypy/messages.py", line 512, in has_no_attr
  File "mypy/messages.py", line 289, in fail
  File "mypy/messages.py", line 264, in report
  File "mypy/errors.py", line 463, in report
  File "mypy/errors.py", line 529, in add_error_info
  File "mypy/errors.py", line 466, in _add_error_info
AssertionError
```

I'm running `dmypy run`, so it's running on the whole code base.  It seems like it happens after I've run dmypy twice. If I run it a third time, it crashes. I can restart dmypy (via `dmypy run`), it works fine once, it works fine if I run `dmypy run` a second time (very quickly as I would expect it's cached a ton of stuff), but then crashes on the third run.

Here's the code base: https://github.com/gerner/stellarpunk/tree/668ea4996ea17bb1da64ea80192c825074f74435
although, I made some edits in my workspace on top of that. I believe if you make any errors mypy will catch you can repro the problem.

Sorry I don't have a minimal repro case, but I'm happy to help out. I'm just not familiar with MyPy internals enough to quickly whip up a minimal repro.

--- a/mypy/dmypy_server.py
+++ b/mypy/dmypy_server.py
@@ -620,6 +620,9 @@ class Server:
         t1 = time.time()
         manager.log(f"fine-grained increment: find_changed: {t1 - t0:.3f}s")
 
+        # Track all modules encountered so far. New entries for all dependencies
+        # are added below by other module finding methods below. All dependencies
+        # in graph but not in `seen` are considered deleted at the end of this method.
         seen = {source.module for source in sources}
 
         # Find changed modules reachable from roots (or in roots) already in graph.
@@ -736,7 +739,9 @@ class Server:
         Args:
             roots: modules where to start search from
             graph: module graph to use for the search
-            seen: modules we've seen before that won't be visited (mutated here!!)
+            seen: modules we've seen before that won't be visited (mutated here!!).
+                  Needed to accumulate all modules encountered during update and remove
+                  everything that no longer exists.
             changed_paths: which paths have changed (stop search here and return any found)
 
         Return (encountered reachable changed modules,
@@ -776,7 +781,9 @@ class Server:
         """Find suppressed modules that have been added (and not included in seen).
 
         Args:
-            seen: reachable modules we've seen before (mutated here!!)
+            seen: reachable modules we've seen before (mutated here!!).
+                  Needed to accumulate all modules encountered during update and remove
+                  everything that no longer exists.
 
         Return suppressed, added modules.
         """

--- a/mypy/dmypy_server.py
+++ b/mypy/dmypy_server.py
@@ -756,7 +756,8 @@ class Server:
                 changed.append((nxt.module, nxt.path))
             elif nxt.module in graph:
                 state = graph[nxt.module]
-                for dep in state.dependencies:
+                ancestors = state.ancestors or []
+                for dep in state.dependencies + ancestors:
                     if dep not in seen:
                         seen.add(dep)
                         worklist.append(BuildSource(graph[dep].path, graph[dep].id, followed=True))

--- a/test-data/unit/daemon.test
+++ b/test-data/unit/daemon.test
@@ -647,3 +647,21 @@ b: str
 from demo.test import a
 [file demo/test.py]
 a: int
+
+[case testDaemonImportAncestors]
+$ dmypy run test.py
+Daemon started
+test.py:2: error: Unsupported operand types for + ("int" and "str")  [operator]
+Found 1 error in 1 file (checked 1 source file)
+== Return code: 1
+$ dmypy run test.py
+test.py:2: error: Unsupported operand types for + ("int" and "str")  [operator]
+Found 1 error in 1 file (checked 1 source file)
+== Return code: 1
+$ dmypy run test.py
+test.py:2: error: Unsupported operand types for + ("int" and "str")  [operator]
+Found 1 error in 1 file (checked 1 source file)
+== Return code: 1
+[file test.py]
+from xml.etree.ElementTree import Element
+1 + 'a'

